### DISCUSSION TITLE
Subscribe to AWS broke with keys?

### DISCUSSION THREAD

**Discussion by spmatthews03** [davidepianca98/KMQTT#38] [Q&A]
My existing broker has authentication to be accesskey,secretkey and token... how might I implement this library to connect?

**Comment by spmatthews03** [comment]
from the looks of it, I am not seeing a possible way to connect to AWS MQTT

**Comment by davidepianca98** [comment]
Please let me know if this is helpful: https://github.com/davidepianca98/KMQTT/issues/24#issuecomment-1727145737

**Reply by spmatthews03** [reply]
Thanks for the reply, unfortunately that doesn't work with our particular use case. Our AWS is immutable, and requires the Sigv4 creds. (accessKey and secretKey)

**Reply by davidepianca98** [reply]
I see, I think you should be able to use this library to generate the URL https://github.com/pedrohbtp/aws_signv4_mqtt. Then you can use KMQTT like this (I'm using the example URL present in the Readme of the linked repo):

```kotlin
val client = MQTTClient(
    MQTTVersion.MQTT3_1_1,
    "<your_iot>.iot.us-east-1.amazonaws.com",
    443,
    TLSClientSettings(serverCertificatePath = "aws.crt"),
    60,
    "/mqtt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=%3Caws_access_key%3E%2F20190216%2Fus-east-1%2Fiotdevicegateway%2Faws4_request&X-Amz-Date=20190216T115653Z&X-Amz-SignedHeaders=host&X-Amz-Signature=92bfc2b5a77489c5786a820044bf6c8ccb3c2f9b40c1d6a5f016b4d9c8662a5e"
) {
    println(it.payload?.toByteArray()?.decodeToString())
}
```

Optionally add username, password and TLS client certificate.

**Reply by spmatthews03** [reply]
This might be a solution! Thanks, which property would that be for? just so I can understand how thats beeing used.
Thanks!

**Reply by davidepianca98** [reply]
In the example of the linked repository the full link `wss://<your_iot>.iot.us-east-1.amazonaws.com/mqtt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=%3Caws_access_key%3E%2F20190216%2Fus-east-1%2Fiotdevicegateway%2Faws4_request&X-Amz-Date=20190216T115653Z&X-Amz-SignedHeaders=host&X-Amz-Signature=92bfc2b5a77489c5786a820044bf6c8ccb3c2f9b40c1d6a5f016b4d9c8662a5e` can be used. Here it has to split in: 
- protocol: `wss://`;
- domain name: `<your_iot>.iot.us-east-1.amazonaws.com`;
- websocket endpoint: `/mqtt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=%3Caws_access_key%3E%2F20190216%2Fus-east-1%2Fiotdevicegateway%2Faws4_request&X-Amz-Date=20190216T115653Z&X-Amz-SignedHeaders=host&X-Amz-Signature=92bfc2b5a77489c5786a820044bf6c8ccb3c2f9b40c1d6a5f016b4d9c8662a5e`.

The last parameter of `MQTTClient` is the websocket path which enables websocket with this specific path.

**Reply by spmatthews03** [reply]
I was still on 0.4.4 I upgraded to 0.4.5 to see the websocket: String param. Thanks. However, I really am unsure how I can successfully test this. There isn't anything where I can see the 'WHY I'm not getting a connection... :/ 
I can do postman and connect with the URL... without TLS.

But using this I just get Connack time out :/ I've stepped through, and I'm not sure anymore what I'm doing wrong. 
I have done as you mentioned...

`client = MQTTClient(
                MQTTVersion.MQTT3_1_1,
                "myaws_domain.iot.eu-west-1.amazonaws.com",
                tls = null,
                port = 443,
                keepAlive = 60,
                webSocket = "/mqtt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AAAAAAAAAAAAAAAAAA%2F20240214%2Feu-west-1%2Fiotdevicegateway%2Faws4_request&X-Amz-Date=20240214T182213Z&X-Amz-SignedHeaders=host&X-Amz-Signature=AAAAAAAAAAAAAAAAAAA&X-Amz-Security-Token=AAAAAAAAAAAAAAAAA"
            )`

**Reply by davidepianca98** [reply]
You cannot just disable TLS and connect to the same port. Also please verify that indeed your broker is on port 443. You are getting connack time out because the server is expecting a TLS packet but receiving an MQTT (actually HTTP for websocket) one, so probably it just doesn't reply to not give out any information about what protocol it is running. There's only so much I can do to help you with this, you first have to make sure you can connect in any way to the broker. You could try following the whole tutorial here https://github.com/pedrohbtp/aws_signv4_mqtt and see if you can connect like the sample.

**Reply by spmatthews03** [reply]
I'm able to connect to the wss://url/mqtt? path I posted above with postman, on 443, without any certs being passed. thats why I'm confused. we're solely in kotlin/android, I'd have to convert the shared python library into kotlin first.

**Reply by davidepianca98** [reply]
As far as I understand you are able to open the TCP connection, but as soon as the MQTT connect packet is sent you get no reply, probably because the server there is expecting a TLS client hello. Let me know if I misunderstood.
What I propose is to test the connection from the computer as shown in the linked repo example ("Testing your url" section of the readme). That example is in Javascript but at least if it works there we know you can at least open the MQTT connection. No need to port anything to Kotlin. If it works on the computer then the parameters are correct and you can start trying this with KMQTT.
Unfortunately AWS uses this non-standard procedure so it's not as simple as any other broker.

**Reply by spmatthews03** [reply]
Hi David, I finally got back around to this. I was able to get the javascript with python working. I was able to connect with wss:// like mentioned above, on port 443 without a passing certs. I migrated to kotlin optimistic that it would work, and unfortunately, the same extact URL, address etc, and kmqtt fails :/ 

`                client = MQTTClient(
                    MQTTVersion.MQTT5,
                    address = host,
                    webSocket = signv4Url,
                    port = 443,
                    tls = null,
                    clientId = generateRandomClientId(),
                   `
                   
where signv4Url is the '/mqtt.....'

**Reply by davidepianca98** [reply]
Hi, in KMQTT you have to specify the `tls` constructor argument of `MQTTClient` to use TLS. I suggest trying with this value
`TLSClientSettings(serverCertificate = "aws.crt")` where aws.crt is this file https://www.amazontrust.com/repository/AmazonRootCA1.pem. In the latest version 0.4.6 you can also pass the content of the file instead of the file path if you prefer (In the same serverCertificate field). Some additional documentation about AWS IoT server authentication: https://docs.aws.amazon.com/iot/latest/developerguide/server-authentication.html. Here it is explained that depending on the endpoint type you might have to use a different cert, but please first try with the linked one.

**Reply by spmatthews03** [reply]
So doing as you suggested I get ...
![image](https://github.com/davidepianca98/KMQTT/assets/10149255/53f7af37-4cbd-40a9-837a-3aa19a6e169a)

I'm familiar with ports and tls, but what I don't understand is why postman, can make the connection on port 443 without certs, and my terminal command line project you linked me to can also make a connection without certs using 443. But for kmqtt, it seems like it absolutely required? 

Any time I put TLSClientSettings in my parameter it immediately erros on runtime with this attached error.

**Reply by davidepianca98** [reply]
Basically the server CA certificate is provided as a means for the client to verify the server. You are right that it is not strictly necessary and if you want I can modify the API such that it is possible to enable TLS without providing the server certificate (I have to verify this can work on every platform though).
The error you are currently seeing is an error from a SUBACK packet. So what happened is that the TLS and MQTT connections were successful, but it seems you are trying to subscribe to a topic which you are not allowed to. Please check everything is correct in your broker permissions and subscribe request and let me know.

**Reply by spmatthews03** [reply]
So thats a good thing, that it seems I'm getting a connection. I'm trying to subscribe to a familiar topic that I have tested in command line and postman and both work... 

unfortunately with that, I'm getting...
![image](https://github.com/davidepianca98/KMQTT/assets/10149255/385c563b-a858-42c7-af81-9786f968fdbc)

But when I remove the subscribe, I don't get any errors. No timeout, CONNACK, or SUBACK errors.

**Reply by davidepianca98** [reply]
Can you please give a brief explanation of how you setup the broker so that I can try to reproduce the issue?

**Reply by spmatthews03** [reply]
I don't have access to the broker. Its a legacy product that was handed off to the contract I'm on, and its essentially a black box. I know it works without the need for certs. and I have 4 other languages connected and subscribed to it. However, unsuccessful here. It seems though that it should work, I just can't figure out what is going on.

**Reply by spmatthews03** [reply]
If you need any more information let me know, I also posted the exact lines for the subscription below.

**Reply by davidepianca98** [reply]
Sorry again for the slow replies. I was able to reproduce the problem with AWS IoT Core in a few instances:
- when the clientId is not set correctly as specified in the policy file of the device
- when the publish/subscribe topic are not set correctly as specified in the policy file of the device
- when the QoS in the subscription or publish is set to EXACTLY_ONCE (this seems to be your case, this is not supported by AWS)

So please try to change the Qos option you are passing in the KMQTT calls to other values like AT_LEAST_ONCE or AT_MOST_ONCE.
Another thing to note is that I have been testing the connection via TCP and not WebSocket like you are using. So on port 8883 and without the websocket settings.
Let me know if changing the Qos works. Otherwise I need more details of the AWS service you are using (if not IoT Core) and the configuration settings you have available.

**Comment by spmatthews03** [comment]
This is my following line.after the creation of the MQTTClient object

client.subscribe(
listOf( Subscription(
"thing-update/GLOBALMARINE/MAOEM/MAOEM_MTC/D-C4DD57AA3AA4/#",
SubscriptionOptions(
Qos.EXACTLY_ONCE))
)
)
client.run()

**Comment by spmatthews03** [comment]
I just tested with AT_MOST_ONCE and AT_LEAST_ONCE, both of those take me
back to UNSPECIFIED_ERROR. where EXACTLY_ONCE gave me a
SocketClosedException.

We are using AWS IOT Core. It has to be wss://<>/mqtt, I get AWS
credentials and I pass it in to a signv4 algorithm and return the URI. I
then use that with the address to make the connection. My topic is set up
correctly, and for our other javascript connections we do not use a
clientId, (I have attempted with both)

I was really hopeful that it was the QoS. And it still might be with a
combination of other things, those of which I have yet to figure out. :/

**Reply by davidepianca98** [reply]
This is what I did to test it with the WebSocket connection and it is working correctly, please let me know if anything is different from your setup.

1. AWS IAM service
2. Create a new user
3. Give it the policy AWSIoTDataAccess
4. Create Access Key
5. Copy the access key and secret key
6. Generate the mqtt websocket URL as mentioned here: https://github.com/pedrohbtp/aws_signv4_mqtt?tab=readme-ov-file#how-to-use-use:
  - pip install boto3
  - pip install aws-signv4-mqtt
  - aws_signv4_mqtt.generate_signv4_mqtt(iot_host, iot_region, access_key, secret_key)
  - Copy the "/mqtt?..." part of the URL

This is my Kotlin code:
```kotlin
val client = MQTTClient(
        MQTTVersion.MQTT5,
        "iot_host.iot.iot_region.amazonaws.com",
        443,
        TLSClientSettings(serverCertificate = "C:\\Users\\david\\IdeaProjects\\KMQTT\\AmazonRootCA1.pem"),
        30,
        webSocket = "/mqtt?..." ) {
    println(it.payload?.toByteArray()?.decodeToString())
}
client.subscribe(listOf(Subscription("sdk/test/js", SubscriptionOptions(Qos.AT_LEAST_ONCE))))
thread {
    while (true) {
        Thread.sleep(5000)
        client.publish(false, Qos.AT_MOST_ONCE, "sdk/test/js", "hello1".toByteArray().toUByteArray())
    }
}
client.run()
```

EXACTLY_ONCE is giving socket closed exception because AWS closes the connection when you try to publish or subscribe with that Qos value. While when subscribing to a topic which is not allowed you get an error in the SUBACK packet saying that the subscription was not enabled but the connection remains open.

Can you send the stack trace of the UNSPECIFIED_ERROR you are getting? It might be because of something else and I would like to see exactly where in the code it is happening.

**Comment by spmatthews03** [comment]
This is exactly what I had done. your steps with the aws_signv4 is what I
had used to test. Then I took the mqtt URI and stuck it into mqttclient. I
then converted that repo into kotlin and have that working in creating the
signed URL which I tested by usin it back into that aws_signv4 repository.
Here is my code...

val signv4Url: String = generateSignv4Mqtt(host, region, awsKeys.accessKey,
awsKeys.secretKey, awsKeys.sessionToken)


val client: MQTTClient = MQTTClient(
MQTTVersion.MQTT5,
address = host,
webSocket = signv4Url,
port = 443,
tls = TLSClientSettings(),
clientId = generateRandomClientId(),
) {
val payload = it.payload?.toByteArray()?.decodeToString()

println("AWSListener: $payload")
}

client.subscribe(
listOf( Subscription(
"thing-update/GLOBALMARINE/MAOEM/MAOEM_MTC/D-C4DD57AA3AA4/#",
SubscriptionOptions(
Qos.*AT_LEAST_ONCE*))
)
)
client.run()

StackTrace
:47.195 13946-14027 AndroidRuntime          com.dometic.test.mtcv2
      E  FATAL EXCEPTION: DefaultDispatcher-worker-3

                        Process: com.dometic.test.mtcv2, PID: 13946

                        mqtt.MQTTException: UNSPECIFIED_ERROR

                        at MQTTClient.handleSuback(MQTTClient.kt:565)

                        at MQTTClient.handlePacket(MQTTClient.kt:389)

                        at MQTTClient.check(MQTTClient.kt:300)

                        at MQTTClient.step(MQTTClient.kt:367)

                        at MQTTClient.run(MQTTClient.kt:377)

                        at
DometicMQTTClient$1.invokeSuspend(DometicMQTTClient.kt:113)

                        at
kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)

                        at
kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)

                        at
kotlinx.coroutines.internal.LimitedDispatcher$Worker.run(LimitedDispatcher.kt:115)

                        at
kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:103)

                        at
kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)

                        at
kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)

                        at
kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)

                        at
kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)

                        Suppressed:
kotlinx.coroutines.internal.DiagnosticCoroutineContextException:
***@***.***, Dispatchers.IO]

**Reply by davidepianca98** [reply]
Again here the broker is returning unspecified error when trying to subscribe. You said you were able to subscribe with this exact topic filter and qos combination also from other clients? Because it really seems the problem is that this client is not allowed to subscribe to this but the broker is giving back a generic error unfortunately. Please try this exact subscription from another client if possible. Also you can try using qos AT_MOST_ONCE and/or going more specific with the topic subscription instead of using the wildcard.

This is getting really weird.

**Comment by spmatthews03** [comment]
Here I'm connected through Postman.
[image: image.png]

**Reply by davidepianca98** [reply]
The image is not showing. Anyway what happens if you try to subscribe to a more specific topic that you know is being used like "thing-update/GLOBALMARINE/MAOEM/MAOEM_MTC/D-C4DD57AA3AA4/_topic1_"?

I also have been testing a bit more with policies. When subscribing to a topic which is not allowed I'm getting NOT_AUTHORIZED. When publishing to a topic not allowed I'm getting socket closed. So this UNSPECIFIED_ERROR for the corresponding subscription in the SUBACK packet is even more baffling.

**Comment by spmatthews03** [comment]
I will give a go at one of the topics instead of wild card, but AWS IoT
says wildcards should work. This really isn't making any sense... I'm happy
the connection is working, and we're getting an acknowledgement. but It is
bizarre that the subscriptions are dying. I don't ever get Not_Authorized.
but I'm not currently publishing. I will be in the future. I'll attempt to
publish something in the meantime too, but I won't get to it for a couple
of days. I appreciate you looking at this, and any more info that you find
out would be greatly helpful.

**Reply by davidepianca98** [reply]
Yes the wildcards should work and indeed in my case they work. But this is so weird I think we should try to simplify the setup as much as possible to try to see what triggers the error.
Yes please try publishing without subscribing and let me know what happens. I will continue trying things in the meantime, thanks for your patience.
